dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/validator/relationship/ConstraintEntityValidator.java

```diff
@@ -0,0 +1,178 @@
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.tracker.validation.validator.relationship;
+
+import static org.hisp.dhis.tracker.validation.ValidationCode.E4012;
+import static org.hisp.dhis.tracker.validation.validator.ValidationUtils.enrollmentExist;
+import static org.hisp.dhis.tracker.validation.validator.ValidationUtils.eventExist;
+import static org.hisp.dhis.tracker.validation.validator.ValidationUtils.trackedEntityInstanceExist;
+import static org.hisp.dhis.tracker.validation.validator.relationship.ValidationUtils.relationshipItemValueType;
+
+import java.util.Optional;
+
+import org.hisp.dhis.relationship.RelationshipConstraint;
+import org.hisp.dhis.relationship.RelationshipType;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.tracker.TrackerImportStrategy;
+import org.hisp.dhis.tracker.TrackerType;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.MetadataIdentifier;
+import org.hisp.dhis.tracker.domain.Relationship;
+import org.hisp.dhis.tracker.domain.RelationshipItem;
+import org.hisp.dhis.tracker.domain.TrackedEntity;
+import org.hisp.dhis.tracker.validation.Reporter;
+import org.hisp.dhis.tracker.validation.ValidationCode;
+import org.hisp.dhis.tracker.validation.Validator;
+
+public class ConstraintEntityValidator implements Validator<Relationship>
+{
+    @Override
+    public void validate( Reporter reporter, TrackerBundle bundle, Relationship relationship )
+    {
+        RelationshipType relationshipType = bundle.getPreheat()
+            .getRelationshipType( relationship.getRelationshipType() );
+
+        validateRelationshipConstraintEntity( reporter, bundle, relationship, "from", relationship.getFrom(),
+            relationshipType.getFromConstraint() );
+        validateRelationshipConstraintEntity( reporter, bundle, relationship, "to", relationship.getTo(),
+            relationshipType.getToConstraint() );
+    }
+
+    private void validateRelationshipConstraintEntity( Reporter reporter, TrackerBundle bundle,
```

Comment:
```suggestion
    private void validateRelationshipConstraint( Reporter reporter, TrackerBundle bundle,
```

The `Entity` in the class and method name is not needed IMHO. We are validating that the constraint is met.

Reply:
No worries 😄 `RelationshipConstraint` defines what type from/to are (specifically the `RelationshipEntity`). It also defines that the trackedEntity needs to be of a specific tracked entity type which we validate in here 

`if ( !type.isEqualTo( constraint.getTrackedEntityType() ) )`

so to me this Validator is validating the constraint and hot just the `constraint.RelationshipEntity`.